metamask: only detectEthereumProvider() once

mrnerdhair opened the pull request:
This follows the style from the new `hdwallet-xdefi` and makes fixing https://github.com/shapeshift/hdwallet/issues/416 easier.

mrnerdhair commented on integration/src/integration.ts:
```diff
@@ -21,10 +21,6 @@ import { terraTests } from "./terra";
 import { kavaTests } from "./kava";
 import { WalletSuite } from "./wallets/suite";
 
-import { ethereum } from "./wallets/mocks/@metamask/detect-provider";
-
-jest.mock("@metamask/detect-provider", () => async () => Promise.resolve(ethereum));
-
```
This was mocking `@metamask/detect-provider` during every wallet's integration test, not just MM.

mrnerdhair commented on integration/src/wallets/__mocks__/@metamask/detect-provider.ts:
```diff
@@ -0,0 +1,24 @@
+export default async function detectEthereumProvider() {
+  return {
+    request: jest.fn(({ method, params }: any) => {
+      switch (method) {
+        case "eth_accounts":
+        case "eth_requestAccounts":
+          return ["0x3f2329C9ADFbcCd9A84f52c906E936A42dA18CB8"];
+        case "personal_sign":
+          const [message] = params;
+
+          if (message === '48656c6c6f20576f726c64')
+            return '0x29f7212ecc1c76cea81174af267b67506f754ea8c73f144afa900a0d85b24b21319621aeb062903e856352f38305710190869c3ce5a1425d65ef4fa558d0fc251b'
+
+          throw new Error('unknown message');
+        case "eth_sendTransaction":
+          const [{ to }] = params;
+
+          return `txHash-${to}`;
+        default:
+          throw new Error(`ethereum: Unkown method ${method}`);
+      }
+    })
+  }
+}
```
This is moved into the correct directory for jest's automock to work.

mrnerdhair commented on integration/src/wallets/metamask.ts:
```diff
@@ -10,8 +12,9 @@ export function createInfo(): core.HDWalletInfo {
 export async function createWallet(): Promise<core.HDWallet> {
-  const wallet = new metamask.MetaMaskHDWallet();
-  await wallet.initialize();
+  const keyring = new core.Keyring();
+  const adapter = metamask.MetaMaskAdapter.useKeyring(keyring);
+  const wallet = await adapter.pairDevice()
```
Creating the wallet via the adapter (which is the public interface) now causes the provider to be pulled in from the `@metamask/detect-provider` mock.

pastaghost commented on packages/hdwallet-metamask/src/metamask.ts:
```diff
@@ -77,14 +76,11 @@ export class MetaMaskHDWallet implements core.HDWallet, core.ETHWallet {
-    return Promise.resolve();
+  public initialize(): never;
+  public initialize(provider: unknown): Promise<any>;
+  public async initialize(provider?: unknown): Promise<any> {
```
If `provider` is a required argument, what's the motivation to mark it as optional in the function definition?

mrnerdhair replied:
```suggestion
  public async initialize(provider?: unknown): Promise<any> {
    if (this.provider) return;
```

mrnerdhair commented:
@cjthompson @pastaghost I've updated the implementation here based on chris's feedback that web's using `wallet.initialize()`. Also updated #423 to harmonize xdefi with this approach: https://github.com/shapeshift/hdwallet/pull/423/commits/cad67f970373cd435e7eb2249da2852e75da3eb4
